Verify `pkg/addr`

henriman opened the pull request:
Still contains workarounds for bugs:

- ([832](https://github.com/viperproject/gobra/issues/832)) String of low byte slice is low; see LowSliceImpliesLowString in verification/utils/sif/assumptions.gobra

TODOs before merge:

- Rebase to `sif-io-spec`
- Open issue to keep track of workarounds

jcp19 commented on pkg/addr/fmt.go:
```diff
@@ -112,6 +112,7 @@ func FormatAS(as_ AS, opts ...FormatOption) string {
 }
 
 // @ requires as_.inRange()
+// @ requires low(as_) && low(sep)
```
why do we mark the inputs as low but not the outputs? it looks like we are introducing a restriction without really benefiting from it

jcp19 replied:
Ah, I get it. I guess you are adding enough annotations so that we pass the low control flow requirement

henriman replied:
Exactly. Though I now realize that `low(sep)` is not necessary anymore (remnant from when I *did* ensure that the output is low)

jcp19 commented on pkg/addr/host.go:
```diff
@@ -475,7 +499,9 @@ func HostFromRaw(b []byte, htype HostAddrType) (res HostAddr, err error) {
 // @ requires len(ip) == HostLenIPv4 || len(ip) == HostLenIPv6
-// @ ensures res.Mem()
+// @ requires low(len(ip)) && (len(ip) == HostLenIPv6 ==> 
+// @ 	low(net.isZeros(ip[0:10])) && low(ip[10] == 255) && low(ip[11] == 255))
```
can we just make the contents of the ip low, rather than the full conditions? I feel like this leaks too many implementation details

henriman replied:
I have done so now, though I still only require that the contents of `ip` are low if `len(ip) == HostLenIPv6`, as otherwise I would have to add sensitivity annotations to `HostAddr.IP` etc. due to `HostIPv4.IP` calling `IP.To4`. Let me know if this is still too specific/implementation-dependent

henriman replied:
I have removed this restriction now

jcp19 commented on verification/dependencies/net/ip.gobra:
```diff
@@ -61,8 +61,12 @@ preserves forall i int :: { &ip[i] } 0 <= i && i < len(ip) ==> acc(&ip[i], R15)
 // To4 converts the IPv4 address ip to a 4-byte representation.
-preserves wildcard ==> forall i int :: { &ip[i] } 0 <= i && i < len(ip) ==> acc(&ip[i], _)
-preserves !wildcard ==> forall i int :: { &ip[i] } 0 <= i && i < len(ip) ==> acc(&ip[i], R20)
+requires   wildcard ==> forall i int :: { &ip[i] } 0 <= i && i < len(ip) ==> acc(&ip[i], _)
+requires  !wildcard ==> forall i int :: { &ip[i] } 0 <= i && i < len(ip) ==> acc(&ip[i], R20)
+requires  low(len(ip)) && (len(ip) == IPv6len ==> 
+	low(isZeros(ip[0:10])) && low(ip[10] == 255) && low(ip[11] == 255))
+ensures    wildcard ==> forall i int :: { &ip[i] } 0 <= i && i < len(ip) ==> acc(&ip[i], _)
+ensures   !wildcard ==> forall i int :: { &ip[i] } 0 <= i && i < len(ip) ==> acc(&ip[i], R20)
```
here, I would probably add the lowness annotation only on the ip contents too

jcp19 commented on verification/dependencies/strings/strings.gobra:
```diff
@@ -158,7 +160,7 @@ func TrimPrefix(s, prefix string) string
 // TrimSuffix returns s without the provided trailing suffix string.
 // If s doesn't end with suffix, s is returned unchanged.
 decreases _
-func TrimSuffix(s, suffix string) string
+pure func TrimSuffix(s, suffix string) string
```
why did you make this pure?

henriman replied:
I marked it as pure to get `low(s) && low(suffix) ==> low(<result>)`, which is necessary for `HostSVCFromString` in pkg/addr/host.go (we branch on the result of `TrimSuffix`, cf. `switch str { ... }`). 

I thought this way would be cleaner than adding the respective postcondition to `TrimSuffix`.

jcp19 replied:
I would prefer adding the explicit contract, as otherwise it may not be clear why this is marked as pure

henriman commented:
@jcp19: I have addressed all your feedback now. The removal of the restriction `len(ip) == HostLenIPv6 ==> ...` on the lowness of `ip` necessitated adding `IsLow` to the precondition of `HostAddr.Pack` and `HostAddr.IP`. I also added `&& low(i) ==>` to quantified assertions where appropriate. (In PR #10, I also already replaced most of these quantified assertions with calls to the corresponding `hyper` functions.)

jcp19 commented on pkg/addr/host.go:
```diff
@@ -92,6 +92,14 @@ const (
+	// NOTE[henri]: I rewrote this comment to be more accurate
+	// Return whether all the underlying data that needs to be low for the
+	// computation of `Pack`, `IP`, `String` is low.
+	//@ ghost
+	//@ requires Mem()
+	//@ decreases
+	//@ pure IsLow() bool
```
this should already be marked as a hyperfunction to guarantee our assumptions make sense

jcp19 replied:
Instead, please add a TODO here so that we make this a hyper function once issue #955 is fixed

jcp19 reviewed (approved):
LGTM, please look at my comment
